### PR TITLE
Migrate from `chan_sip` module to `res_pjsip`

### PR DESCRIPTION
**Pull request by nanosonde**
This PR addresses issue #17 and migrates the generated SIP endpoints from chan_sip (sip.conf) to res_pjsip (pjsip.conf).

The PR assumes that the same password (auto add secret variable) is used for all generated endpoints.
The username is fixed to "sipjs".

### CODE REVIEW

**Comment by felipecrs** [comment]
This looks neat!

File: asterisk/rootfs/usr/share/tempio/pjsip_default.conf.gtpl [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+; For additional parameters go to https://wiki.asterisk.org/wiki/display/AST/Asterisk+18+Configuration_res_pjsip
+
+{{ if .auto_add }}
+{{ $secret := .auto_add_secret }}
+
+; Common AUTH parameters (same username and same secret for all auto-generated endpoints)
+[sipjs-phone-auth]
+type=auth
+auth_type=userpass
+username=sipjs
```
Comment (TECH7Fox) [inline asterisk/rootfs/usr/share/tempio/pjsip_default.conf.gtpl:13]: Maybe change the username to the person name? Like the callerid.

**Review by TECH7Fox — CHANGES_REQUESTED** [review]
Looks great! Please change the branch to develop and if it's tested it's ready to merge.

**Comment by TECH7Fox** [comment]
Quick question, is it now possible to register multiple times on the same endpoint? I saw the `max_contacts` setting.

That would allow it to open the HA-SIP card on multiple devices.

**Comment by pergolafabio** [comment]
I tried max contact von FreePBX before, but never got it too work, the first registration had an incoming call, but aborted after a sec.... While the second one was ringing...

Hopefully it works on asterisk

**Comment by nanosonde** [comment]
Hi!

Yes, this should be possible. However, I have never tried it.
Important is that the dialplan uses the correct way to dial all contacts that are currently registered by using the example described here: 
https://wiki.asterisk.org/wiki/display/AST/Dialing+PJSIP+Channels

PJSIP_DIAL_CONTACTS has to be used.

https://wiki.asterisk.org/wiki/display/AST/Asterisk+18+Function_PJSIP_DIAL_CONTACTS

**Comment by pergolafabio** [comment]
Aha, I didn't try that, was just using FreePBX, no custom dialplan...

Hopefully it uses only 1 channel, then it's maybe possible to send also video in early media stage (session 183), that would be great!!

That was always failing when calling to queue/groups

**Comment by nanosonde** [comment]
> Looks great! Please change the branch to develop and if it's tested it's ready to merge.

Do you want to merge main into develop?
It is behind.

File: asterisk/rootfs/usr/share/tempio/pjsip_default.conf.gtpl (hunk 1 above)
Comment (nanosonde) [reply asterisk/rootfs/usr/share/tempio/pjsip_default.conf.gtpl:13]: Makes sense to better distinguish between them.

BTW: where does the person name come from? I have added user account (which has a person associated) in HA with only my first name. However, the callerid also contains my lastname. Never entered it in HA.

Comment (felipecrs) [reply asterisk/rootfs/usr/share/tempio/pjsip_default.conf.gtpl:13]: @nanosonde it comes from HA Persons API (how your last name is there, Idk).

**Comment by felipecrs** [comment]
Is this ready to merge?

**Comment by felipecrs** [comment]
I suggest to force rebase the PR against the updated develop branch to avoid all these commits to be accounted.

**Comment by pergolafabio** [comment]
just wondering, i saw removing chan_sip in the PR, can you leave it for a while? so we have time to migrate once it get released? :-)

**Comment by felipecrs** [comment]
> just wondering, i saw removing chan_sip in the PR, can you leave it for a while? so we have time to migrate once it get released? :-)

@pergolafabio I suggest you simply do not upgrade the add-on, if you want to keep using chan_sip.

**Comment by felipecrs** [comment]
Anyway, the files at /config/asterisk won't be deleted, so you will have your sip_custom.conf there for reference.

**Comment by nanosonde** [comment]
We cannot have chan_sip and res_pjsip working at the same time on the same ports.
So this PR would disable the chan_sip module loading only. If you want to change behaviour you are free to do so and change the config manually.

**Comment by pergolafabio** [comment]
ah ok, its as easy changing this line?

noload => chan_sip.so

**Comment by nanosonde** [comment]
@felipecrs Shall we keep generating sip_default.conf at the same time with pjsip_default.conf from the templates as some kind of convenience for those who want to use chan_sip?

**Comment by pergolafabio** [comment]
isnt it possible like in freepbx?
chansip 5160
pjsip 5060 ?

**Comment by felipecrs** [comment]
> @felipecrs Shall we keep generating sip_default.conf at the same time with pjsip_default.conf from the templates as some kind of convenience for those who want to use chan_sip?

If it's for backwards compatibility purposes, no. This add-on is in very early stages still, such breaking changes can happen. 

I think bumping the major version and writing a good changelog is enough.

Anyway, this is my personal suggestion. @TECH7Fox what do you think?

**Comment by nanosonde** [comment]
> Anyway, this is my personal suggestion. @TECH7Fox what do you think?

I think we should keep the generation and mark it as DEPRECATED. After pjsip has proven to work we could easily remove the old chan_sip stuff.

**Comment by pergolafabio** [comment]
yeah, just make it clear i changelog! :-)

go ahead, thats my opnion too, dont think so many users are using it already

**Comment by TECH7Fox** [comment]
I think we can just switch to pjsip and remove the old sip generation. You register the same right? So the card config will be the same.

People that use custom sip devices do need to change to pjsip. We should make a wiki page to explain how to do so.

And the dialplan. People will have to delete the old file or change it themself.

**Comment by nanosonde** [comment]
@TECH7Fox 
If the card config shall stay the same, we have to change again the `username`. 
Currently I have set it to the person name.

Shall I change it to the extension number? e.g. 100, 101, etc....

**Comment by nanosonde** [comment]
BTW: I am currently testing on my productin machine.... stay tuned...

**Comment by nanosonde** [comment]
The qualify functionality in pjsip seems be broken somehow for websocket connections.
See this: https://community.asterisk.org/t/websocket-fail-with-pjsip-eunsuptransport/89683/9

I also get this error for our generated endpoints:
` res_pjsip.c:4828 endpt_send_request: Error 171060 ‘Unsupported transport (PJSIP_EUNSUPTRANSPORT)’ sending OPTIONS request to endpoint 100`

However, it is notthat serious as websocket connection are based on the connection-oriented TCP protocol so asterisk will immediately set the endpoint to be unreachable if the websocket connection gets closed.

**Comment by nanosonde** [comment]
Just a small sign of life...

I have spent the whole time now to find out why asterisk 18.2.2. and pjsip is not working properly.
Unfortunately, I have no exact reason what the root cause is, but it is related to the DNS lookup and transports.
When dialing an endpoint, I always got error 503: congestion.
After setting all debug level to maximum, I realized that something goes wrong when asterisk tries to lookup the IP of the rewritten CONTACT header from the registered endpoint. It only had the bare IP address and was not able to lookup the transport for this. Of course, it is an IP address and not a domain with SRV records which might contain an indicitation of what transport to use. 

After experimenting a lot, I finally moved this addon from Alpine-based HASS image to Debian-based HASS image.
Asterisk 18.10.1 is now compiled on our own. 
What shall I say? With the same config files it is now working.....

So I will definitly go with the Debian-based HASS base images and manually compile asterisk.

How shall we proceed?

**Comment by nanosonde** [comment]
BTW: I found other issues while searching the web related to Asterisk/Alpine combo.
So we should really move away from this.

**Comment by felipecrs** [comment]
> How shall we proceed?

How about we do the migration to Debian in another PR, and then we come back here and just rebase this one once the other PR is merged?

**Comment by nanosonde** [comment]
Some good news: I have just tested the simultaneous PJSIP channel dialing to all registered contacts for the same user. 
Seems to work. My browser showed the incoming call and also my HA companion app on my iPhone. 
I can then pickup the call from either one.

**Comment by pergolafabio** [comment]
> Some good news: I have just tested the simultaneous PJSIP channel dialing to all registered contacts for the same user. Seems to work. My browser showed the incoming call and also my HA companion app on my iPhone. I can then pickup the call from either one.
> 
> The testing already took place on the debian based addon with manually compiled Asterisk 18.10.1.

good jobs!! do you know if you use "max contacts", if early media is also passed to all contacts? (session 183) ?

**Comment by nanosonde** [comment]
> The qualify functionality in pjsip seems be broken somehow for websocket connections. See this: https://community.asterisk.org/t/websocket-fail-with-pjsip-eunsuptransport/89683/9
> 
> I also get this error for our generated endpoints: ` res_pjsip.c:4828 endpt_send_request: Error 171060 ‘Unsupported transport (PJSIP_EUNSUPTRANSPORT)’ sending OPTIONS request to endpoint 100`
> 
> However, it is notthat serious as websocket connection are based on the connection-oriented TCP protocol so asterisk will immediately set the endpoint to be unreachable if the websocket connection gets closed.

BTW This issue was also solved by the change to Debian-based image and asterisk 18.10.1. My educated guess is that the root cause is related.

**Comment by nanosonde** [comment]
BTW: concering simultaneous PJSIP channel dialing: 
This even works if you disable "HIDE ME" in the card and then call yourself. All other UAs will start ringing and you can answer and have a call.

Early media and progress (183) have to be checked though.

**Comment by pergolafabio** [comment]
would be wonderfull

**Comment by felipecrs** [comment]
Wondering if:

```
menuselect/menuselect --disable-category MENUSELECT_CHANNELS --enable chan_pjsip
```

Helps us to save some space.

**Comment by nanosonde** [comment]
> Wondering if:
> 
> ```
> menuselect/menuselect --disable-category MENUSELECT_CHANNELS --enable chan_pjsip
> ```
> 
> Helps us to save some space. Supposedly it will cause Asterisk to build only chan_pjsip.

Yes, but maybe one or the other is also still required. Not sure.
Don't optimize to early. ;-)

**Comment by felipecrs** [comment]
> Don't optimize too early. ;-)

@nanosonde yes yes. I just pasted here for future reference. :)

**Comment by nanosonde** [comment]
So I will rebase this to the `develop` branch, right?

**Comment by felipecrs** [comment]
@nanosonde yes. Now, I'm wondering if we should keep building `chan_sip`. I think we don't need, as it's deprecated anyway.

@TECH7Fox WDYT?

**Comment by TECH7Fox** [comment]
> Now, I'm wondering if we should keep building chan_sip.

Maybe some people still want to use it. Then at least the option is still there right? If it doesn't take too much space and build time I think we should just keep it.

But what do you guys think?

**Comment by TECH7Fox** [comment]
Btw, some builds are taking > 30 min and others 5 minutes, which I do recognize from before. Do you know the reason for this? Could we maybe fix this?

**Comment by nanosonde** [comment]
If we keep chan_sip then there is a danger of people trying to find snippets for old chan_sip stuff and try to enable it and probably break things. 😉

**Comment by nanosonde** [comment]
The build for the ARM architecture are slow because they use qemu-arm.

This is the case also on my local dev machine.

**Comment by TECH7Fox** [comment]
> The builds for the ARM architectures are slow because they use qemu-arm.

Ah, alright. Thanks for the explanation. :)

**Comment by felipecrs** [comment]
Yes, we can implement cache. But the current Dockerfile isn't very cache friendly, we should ideally pin the version of all the apt-get dependencies.

Anyway, let's discuss this in another issue.

**Comment by felipecrs** [comment]
#119

**Comment by felipecrs** [comment]
> If we keep chan_sip then there is a danger of people trying to find snippets for old chan_sip stuff and try to enable it and probably break things. 😉

That's true. And I see no reason to support people who would like to use `chan_sip`. They can keep the old version of the add-on if they want.

File: asterisk/rootfs/etc/asterisk/modules.conf [hunk 2]
```diff
@@ -1,8 +1,5 @@
 [modules]
 autoload=yes
-noload => chan_sip.so
-noload => chan_alsa.so
-noload => chan_console.so
```
Comment (felipecrs) [inline asterisk/rootfs/etc/asterisk/modules.conf:4]: I know nothing about this, so... if it's okay for you and @TECH7Fox, I'm good too.

File: asterisk/rootfs/usr/share/tempio/pjsip_default.conf.gtpl [hunk 3]
```diff
@@ -15,31 +15,30 @@ password={{ $secret }}
...
 type=endpoint
 send_rpid=yes
 send_pai=yes
 device_state_busy_at=1
-dtls_auto_generate_cert=yes
 webrtc=yes
 ; Setting webrtc=yes is a shortcut for setting the following options:
 ; use_avpf=yes
 ; media_encryption=dtls
+; dtls_auto_generate_cert=yes (if dtls_cert_file is not set)
```
Comment (felipecrs) [inline asterisk/rootfs/usr/share/tempio/pjsip_default.conf.gtpl:32]: wow this one is nice!

File: asterisk/rootfs/etc/asterisk/modules.conf (hunk 2 above)
Comment (nanosonde) [reply asterisk/rootfs/etc/asterisk/modules.conf]: ihave disabled those modules during build. We do not have any sound output using ALSA HW.
Same goes for chan_console which uses portaudio to output directly to HW.
chan_sip will be disbaled during build with this PR.

File: asterisk/rootfs/usr/share/tempio/pjsip_default.conf.gtpl [hunk 4]
```diff
@@ -15,31 +15,30 @@ password={{ $secret }}
...
 ; media_use_received_transport=yes
 ; rtcp_mux=yes
 context=default
 disallow=all
-allow=ulaw,alaw,speex,gsm,g726,g723,opus
-{{ if .video_support }}
-allow=h263,h263p,h264,vp8
-{{ end }}
+allow=ulaw,alaw,speex,gsm,g726,g723,g722,opus
+{{ if .video_support }}allow=h264,vp8,vp9{{ end }}
```
Comment (felipecrs) [inline asterisk/rootfs/usr/share/tempio/pjsip_default.conf.gtpl:41]:
```
{{ if .video_support -}}
allow=h263,h263p,h264,vp8
{{- end }}
```

Do it like this instead. Notice the hyphens.

**Review by felipecrs — CHANGES_REQUESTED** [review]

File: asterisk/rootfs/usr/share/tempio/pjsip_default.conf.gtpl (hunk 3 above)
Comment (nanosonde) [reply asterisk/rootfs/usr/share/tempio/pjsip_default.conf.gtpl:32]: Yes, see config docs here: https://wiki.asterisk.org/wiki/display/AST/Asterisk+18+Configuration_res_pjsip#Asterisk18Configuration_res_pjsip-endpoint_dtls_auto_generate_cert

**Review by TECH7Fox — APPROVED** [review]

**Review by felipecrs — APPROVED** [review]

**Comment by nanosonde** [comment]
> > Some good news: I have just tested the simultaneous PJSIP channel dialing to all registered contacts for the same user. Seems to work. My browser showed the incoming call and also my HA companion app on my iPhone. I can then pickup the call from either one.
> > The testing already took place on the debian based addon with manually compiled Asterisk 18.10.1.
> 
> good jobs!! do you know if you use "max contacts", if early media is also passed to all contacts? (session 183) ?

@pergolafabio 
I think you have also been interested in early media support of asterisk and on how to do that.

Unfortunately, it is not possibel with more than one client at the same time. Call groups are also not possible.
The Asterisk devs also seem to have also no interest in implemeting this. For them, this is uncommon stuff.
See here: https://community.asterisk.org/t/frustrated-cant-get-early-media-to-work/90350/50

**Pull request merged**